daemon/containerd/image_test.go

```diff
@@ -0,0 +1,280 @@
+package containerd
+
+import (
+	"context"
+	"io"
+	"math/rand"
+	"path/filepath"
+	"testing"
+
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/metadata"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/log/logtest"
+	"github.com/distribution/reference"
+	dockerimages "github.com/docker/docker/daemon/images"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+
+	"go.etcd.io/bbolt"
+
+	"gotest.tools/v3/assert"
+	is "gotest.tools/v3/assert/cmp"
+)
+
+func TestLookup(t *testing.T) {
+	ctx := namespaces.WithNamespace(context.TODO(), "testing")
+	ctx = logtest.WithT(ctx, t)
+	mdb := newTestDB(ctx, t)
+	service := &ImageService{
+		images: metadata.NewImageStore(mdb),
+	}
+
+	testImages := []images.Image{
+		{
+			Name:   "docker.io/library/ubuntu:latest",
+			Target: desc(10),
+		},
+		{
+			Name:   "docker.io/library/ubuntu:latest@" + digestFor(10).String(),
+			Target: desc(10),
+		},
+		{
+			Name:   "docker.io/library/ubuntu:latest@" + digestFor(11).String(),
+			Target: desc(11),
+		},
+		{
+			Name:   "docker.io/library/abcdef:latest",
+			Target: desc(12),
+		},
+		{
+			Name:   "docker.io/library/abcdef:latest@" + digestFor(12).String(),
+			Target: desc(12),
+		},
+		{
+			Name:   "docker.io/library/sha256:defcab",
+			Target: desc(13),
+		},
+		{
+			Name:   "docker.io/test/volatile:retried",
+			Target: desc(14),
+		},
+		{
+			Name:   "docker.io/test/volatile:inconsistent",
+			Target: desc(15),
+		},
+	}
+	for _, img := range testImages {
+		if _, err := service.images.Create(ctx, img); err != nil {
+			t.Fatalf("failed to create image %q: %v", img.Name, err)
+		}
+	}
+
+	for _, tc := range []struct {
+		lookup string
+		img    *images.Image
+		all    []images.Image
+		err    error
+	}{
+		{
+			// Get ubuntu images with default "latest" tag
+			lookup: "ubuntu",
+			img:    &testImages[0],
+			all:    []images.Image{testImages[0], testImages[1]},
```

Comment:
Instead of referencing the test image by their index, could we name them?

```go
ubuntuLatest :=  images.Image{
	Name:   "docker.io/library/ubuntu:latest",
	Target: desc(10),
}
ubuntuLatestDigested :=  images.Image{
	Name:   "docker.io/library/ubuntu:latest"@" + digestFor(10).String(),
	Target: desc(10),
}
testImages := []images.Image{
		ubuntuLatest,
		ubuntuLatestDigested
...
```

```suggestion
			all:    []images.Image{ubuntuLatest, ubuntuLatestDigested},
```